### ISSUE TITLE
Range query with dates not working properly on 2.0.2

### ISSUE THREAD

**Issue by alissonsales** [elastic/elasticsearch#23393]
**Elasticsearch version**: 2.0.2

**Plugins installed**: [head]

**JVM version**:
```
# java -version
openjdk version "1.8.0_111"
OpenJDK Runtime Environment (build 1.8.0_111-8u111-b14-2~bpo8+1-b14)
OpenJDK 64-Bit Server VM (build 25.111-b14, mixed mode)
```

**OS version**: using docker for mac running official docker image elasticsearch:2.0 (adccd7124f26)

**Description of the problem including expected versus actual behavior**:

The range filter using dates without time is not behaving like expected.

The `lte` filter is not converting the date `2017-02-06` to `2017-02-06T23:59:59.999Z`.

I've put on a script to reproduce the issue:

```bash
# test_date_query.sh
base_url=$1

curl "${base_url}"
echo ""
echo "Removing index"
curl -X DELETE "$base_url/my_index"
echo ""

echo "Creating index and mapping"
curl -X POST "${base_url}/my_index" -d '
{
  "settings": { "number_of_shars": "1", "number_of_replicas": "1" },
  "mappings": {
    "user": {
      "properties": {
        "name": { "type": "string" },
        "created_at": {"type": "date", format: "date_optional_time" }
      }
    }
  }
}
'
echo ""
echo "Add documents"
curl -X PUT "${base_url}/my_index/user/_bulk?refresh=true" -d '
{ "index": { "_id": "1"} }
{ "name": "John Doe", "created_at": "2017-02-05T12:30:45.123Z" }
{ "index": { "_id": "2"} }
{ "name": "The Man", "created_at": "2017-02-06T18:21:35.321Z" }
'
echo ""
echo "Query"

curl -X POST "${base_url}/my_index/user/_search" -d '
{
  "explain": true,
  "query": {
    "range": {
      "created_at": {
        "gte": "2017-01-01",
        "lte": "2017-02-06"
      }
    }
  }
}
'
echo ""
```

#### ES 1.7.6

```
bash test_date_query.sh http://localhost:9201
{
  "status" : 200,
  "name" : "Dr. John Grey",
  "cluster_name" : "elasticsearch",
  "version" : {
    "number" : "1.7.6",
    "build_hash" : "c730b59357f8ebc555286794dcd90b3411f517c9",
    "build_timestamp" : "2016-11-18T15:21:16Z",
    "build_snapshot" : false,
    "lucene_version" : "4.10.4"
  },
  "tagline" : "You Know, for Search"
}

Removing index
{"acknowledged":true}
Creating index and mapping
{"acknowledged":true}
Add documents
{"took":199,"errors":false,"items":[{"index":{"_index":"my_index","_type":"user","_id":"1","_version":1,"status":201}},{"index":{"_index":"my_index","_type":"user","_id":"2","_version":1,"status":201}}]}
Query
{"took":6,"timed_out":false,"_shards":{"total":5,"successful":5,"failed":0},"hits":{"total":2,"max_score":1.0,"hits":[{"_shard":2,"_node":"Y8zyM5_rQEeSxwIZN9UOCQ","_index":"my_index","_type":"user","_id":"1","_score":1.0,"_source":{ "name": "John Doe", "created_at": "2017-02-05T12:30:45.123Z" },"_explanation":{"value":1.0,"description":"ConstantScore(created_at:[1483228800000 TO 1486425599999]), product of:","details":[{"value":1.0,"description":"boost"},{"value":1.0,"description":"queryNorm"}]}},{"_shard":3,"_node":"Y8zyM5_rQEeSxwIZN9UOCQ","_index":"my_index","_type":"user","_id":"2","_score":1.0,"_source":{ "name": "The Man", "created_at": "2017-02-06T18:21:35.321Z" },"_explanation":{"value":1.0,"description":"ConstantScore(created_at:[1483228800000 TO 1486425599999]), product of:","details":[{"value":1.0,"description":"boost"},{"value":1.0,"description":"queryNorm"}]}}]}}
```

#### ES 2.0.2

```
bash test_date_query.sh http://localhost:9200
{
  "name" : "Windshear",
  "cluster_name" : "elasticsearch",
  "version" : {
    "number" : "2.0.2",
    "build_hash" : "6abf5d8e5b37c2735d8e2f110f9743c453f71e92",
    "build_timestamp" : "2015-12-16T12:49:58Z",
    "build_snapshot" : false,
    "lucene_version" : "5.2.1"
  },
  "tagline" : "You Know, for Search"
}

Removing index
{"acknowledged":true}
Creating index and mapping
{"acknowledged":true}
Add documents
{"took":290,"errors":false,"items":[{"index":{"_index":"my_index","_type":"user","_id":"1","_version":1,"_shards":{"total":2,"successful":1,"failed":0},"status":201}},{"index":{"_index":"my_index","_type":"user","_id":"2","_version":1,"_shards":{"total":2,"successful":1,"failed":0},"status":201}}]}
Query
{"took":4,"timed_out":false,"_shards":{"total":5,"successful":5,"failed":0},"hits":{"total":1,"max_score":1.0,"hits":[{"_shard":3,"_node":"5q3xORWLTFGhduzSOuCwWg","_index":"my_index","_type":"user","_id":"1","_score":1.0,"_source":{ "name": "John Doe", "created_at": "2017-02-05T12:30:45.123Z" },"_explanation":{"value":1.0,"description":"sum of:","details":[{"value":1.0,"description":"created_at:[1483228800000 TO 1486339200000], product of:","details":[{"value":1.0,"description":"boost","details":[]},{"value":1.0,"description":"queryNorm","details":[]}]},{"value":0.0,"description":"match on required clause, product of:","details":[{"value":0.0,"description":"# clause","details":[]},{"value":1.0,"description":"_type:user, product of:","details":[{"value":1.0,"description":"boost","details":[]},{"value":1.0,"description":"queryNorm","details":[]}]}]}]}}]}}
```

#### ES 2.4.4

```
$ bash test_date_query.sh http://localhost:9202
{
  "name" : "Skull the Slayer",
  "cluster_name" : "elasticsearch",
  "cluster_uuid" : "0DQ9_A2sR5Gqcwqe4dO1lQ",
  "version" : {
    "number" : "2.4.4",
    "build_hash" : "fcbb46dfd45562a9cf00c604b30849a6dec6b017",
    "build_timestamp" : "2017-01-03T11:33:16Z",
    "build_snapshot" : false,
    "lucene_version" : "5.5.2"
  },
  "tagline" : "You Know, for Search"
}

Removing index
{"acknowledged":true}
Creating index and mapping
{"acknowledged":true}
Add documents
{"took":360,"errors":false,"items":[{"index":{"_index":"my_index","_type":"user","_id":"1","_version":1,"_shards":{"total":2,"successful":1,"failed":0},"status":201}},{"index":{"_index":"my_index","_type":"user","_id":"2","_version":1,"_shards":{"total":2,"successful":1,"failed":0},"status":201}}]}
Query
{"took":8,"timed_out":false,"_shards":{"total":5,"successful":5,"failed":0},"hits":{"total":2,"max_score":1.0,"hits":[{"_shard":2,"_node":"t3_6DM8RS3u01Igmn8hpmA","_index":"my_index","_type":"user","_id":"2","_score":1.0,"_source":{ "name": "The Man", "created_at": "2017-02-06T18:21:35.321Z" },"_explanation":{"value":1.0,"description":"sum of:","details":[{"value":1.0,"description":"created_at:[1483228800000 TO 1486425599999], product of:","details":[{"value":1.0,"description":"boost","details":[]},{"value":1.0,"description":"queryNorm","details":[]}]},{"value":0.0,"description":"match on required clause, product of:","details":[{"value":0.0,"description":"# clause","details":[]},{"value":1.0,"description":"_type:user, product of:","details":[{"value":1.0,"description":"boost","details":[]},{"value":1.0,"description":"queryNorm","details":[]}]}]}]}},{"_shard":3,"_node":"t3_6DM8RS3u01Igmn8hpmA","_index":"my_index","_type":"user","_id":"1","_score":1.0,"_source":{ "name": "John Doe", "created_at": "2017-02-05T12:30:45.123Z" },"_explanation":{"value":1.0,"description":"sum of:","details":[{"value":1.0,"description":"created_at:[1483228800000 TO 1486425599999], product of:","details":[{"value":1.0,"description":"boost","details":[]},{"value":1.0,"description":"queryNorm","details":[]}]},{"value":0.0,"description":"match on required clause, product of:","details":[{"value":0.0,"description":"# clause","details":[]},{"value":1.0,"description":"_type:user, product of:","details":[{"value":1.0,"description":"boost","details":[]},{"value":1.0,"description":"queryNorm","details":[]}]}]}]}}]}}
```

The expected number of hits for this query is 2, but ES 2.0.2 is returning only 1.

### Number of documents returned:

1.7.6: 2
2.0.2: 1
2.4.4: 2

### Ranges returned on explain:

1.7.6: [1483228800000 TO 1486425599999]
2.0.2: [1483228800000 TO 1486339200000]
2.4.4: [1483228800000 TO 1486425599999]

Versions 1.7.6 and 2.4.4 are converting the date used on the lte to "end of the day" time

**Comment by dadoonet** [comment]
So it's fixed in 2.4. This is it what you are saying right?
I guess we can close it?

**Comment by alissonsales** [comment]
@dadoonet Yep, it's working properly on 2.4. I thought it was worth reporting since I don't know if you guys advise the upgrade to 2.4 instead.

I've found this issue while analyzing what's required to upgrade from ES 1.7 to 2.0. Do you think it's safe to go straight to 2.4 from 1.7?

**Comment by dadoonet** [comment]
I believe it is. FWIW we won't fix in 2.0 branch

**Comment by alissonsales** [comment]
Cool, thanks @dadoonet
